### PR TITLE
Split language extensions source up with \input

### PR DESCRIPTION
**Pull request by johnwhitington**
The file `ocaml/manual/manual/refman/exten.etex` is too long in two senses:

1. The source .etex is more than 2500 lines
2. It has some content which would be better moved elsewhere, making the page have fewer sections.

This PR addresses (1), so that addressing (2) will become easier. I will have proposals on (2) later.

### CODE REVIEW

File: manual/manual/refman/exten.etex [hunk 1]
```diff
@@ -8,2767 +8,109 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
...
-Another application where the absence of strengthening comes handy, is
-to provide an alternative implementation for an existing module.
-\begin{caml_example*}{verbatim}
-module MySet : module type of Set = struct
-  include Set[@@ellipsis]
-end
-\end{caml_example*}
-This idiom guarantees that "Myset" is compatible with Set, but allows
-it to represent sets internally in a different way.
+\input{recoveringtype.tex}
```
Comment (Octachron) [inline manual/manual/refman/exten.etex]: Is there a reason to change the name compared to the cutname?

File: manual/manual/refman/exten.etex [hunk 2]
```diff
@@ -8,2767 +8,109 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
...
-
-Note that, unlike type declarations, type substitution declarations are not
-recursive, so substitutions like the following are rejected:
-
-\begin{caml_example}{toplevel}
-module type S = sig
-  type 'a poly_list := [ `Cons of 'a * 'a poly_list | `Nil ]
-end [@@expect error];;
-\end{caml_example}
+\input{substitutingsig.tex}
```
Comment (Octachron) [inline manual/manual/refman/exten.etex]: Same as above, `signaturesubstitution` seems fine to me.

File: manual/manual/refman/exten.etex (hunk 1 above)
Comment (johnwhitington) [reply manual/manual/refman/exten.etex]: I will fix these in a moment.

Comment (johnwhitington) [reply manual/manual/refman/exten.etex]: Why do some sections (e.g recursive-modules) not have a cutname anyway?

Comment (Octachron) [reply manual/manual/refman/exten.etex]: They all should have a cutname and indeed the one for recursive modules is missing, leading to an anonymous page name: https://ocaml.org/releases/4.11/htmlman/manual024.html . 
This PR is probably the right time to fix this issue.

**Comment by gasche** [comment]
I have a hard time finding the files I'm looking for in the manual (as a start, I never know in which directory to look at, `refman`, `cmds` etc.; would you look in `cmds/` for the FFI documentation?), and I am a bit worried that adding many files in `refman` will make things worse. Could they go into an `refman/exten` subdirectory?

**Comment by johnwhitington** [comment]
Yes, we can do this.

By the way, does anyone else find `manual/manual` confusing too? Could we change it to another name? Or maybe just squash it, and move everything into `manual/`?

**Comment by Octachron** [comment]
I concur that having an `extensions` subdirectory would make it easier to find the extension related sections.
Another option would be to have a `manual/manual/extensions` subdirectory.

Concerning `manual/manual`, we could maybe use `manual/src` to avoid this repetition and make it clearer that the second level contains the manual source files and the various processed output.

**Comment by gasche** [comment]
(For me `manual/manual` is not an issue because at each step I know where I want to go -- to the manual sources. But I'm happy to let people change to what they prefer.)

File: manual/manual/refman/Makefile [hunk 3]
```diff
@@ -10,7 +10,7 @@ CAMLLATEX = $(SET_LD_PATH) \
 TEXQUOTE = $(OCAMLRUN) $(TOOLS)/texquote2
 TRANSF = $(SET_LD_PATH) $(OCAMLRUN) $(TOOLS)/transf
 
-
+VPATH = extensions
```
Comment (gasche) [inline manual/manual/refman/Makefile]: A simpler approach (in that it does not use fancy features such as `VPATH`) would be to have an EXTENSION_FILES variable with the extensions, and then include `$(addsuffix extension/, $(EXTENSION_FILES))` in `FILES`.

**Comment by johnwhitington** [comment]
(Not sure what the problem is with this last commit. I used `git mv manual/manual manual/src` so I'm not sure why it is no longer mergeable)

**Comment by Octachron** [comment]
Git doesn't really know about moves: `git mv` deletes and recreates the files. Git let the higher-level tool present the changes in a human-friendly way. In this case, there is a conflict because the deleted files are not the same in your branch and trunk (due to the changes in the code example macros).

File: manual/manual/refman/Makefile (hunk 3 above)
Comment (johnwhitington) [reply manual/manual/refman/Makefile]: Done, but with addprefix :-)

**Comment by johnwhitington** [comment]
Hold on here, I appear to have broken manual building...

**Comment by johnwhitington** [comment]
I have restored the four conflicting files in their new place in `manual/src`. Hopefully this means someone with correct permissions can now merge.

**Comment by Octachron** [comment]
Rebasing should work better to remove the conflict on the deleted files here.

**Comment by johnwhitington** [comment]
The flambda failing check ends with this, which I don't understand, but presume is a check not being done in the two passing checks:

```
++ git status --porcelain
578
+ test -z '?? manual/src/cmds/warnings-help.etex
579
?? manual/src/warnings-help.etex'
580
Error: Process completed with exit code 1.
```

**Comment by lthls** [comment]
The test seems to complain that building the manual creates files that are not ignored by git.
The other checks do not build the stdlib manpages, which would explain why they didn't complain.
I think you just need to update the root `.gitignore` file with the new manual path.

**Comment by Octachron** [comment]
Indeed: https://github.com/ocaml/ocaml/blob/trunk/.gitignore#L94

**Comment by johnwhitington** [comment]
Thanks. Done.

**Comment by johnwhitington** [comment]
A little progress. Now we see:

```
++ git status --porcelain
1069
+ test -z ''
1070
++ git ls-files --others -i --exclude-standard
1071
+ test -z 'manual/src/refman/extensions/alerts.tex
1072
manual/src/refman/extensions/alerts_camltex.tex
1073
manual/src/refman/extensions/attributes.tex
1074
manual/src/refman/extensions/attributes_camltex.tex
1075
manual/src/refman/extensions/bigarray.tex
1076
manual/src/refman/extensions/bigarray_camltex.tex
1077
manual/src/refman/extensions/bindingops.tex
1078
manual/src/refman/extensions/bindingops_camltex.tex
1079
manual/src/refman/extensions/doccomments.tex
1080
manual/src/refman/extensions/doccomments_camltex.tex
1081
manual/src/refman/extensions/emptyvariants.tex
1082
manual/src/refman/extensions/emptyvariants_camltex.tex
1083
manual/src/refman/extensions/extensiblevariants.tex
1084
manual/src/refman/extensions/extensiblevariants_camltex.tex
1085
manual/src/refman/extensions/extensionnodes.tex
1086
manual/src/refman/extensions/extensionnodes_camltex.tex
1087
manual/src/refman/extensions/extensionsyntax.tex
1088
manual/src/refman/extensions/extensionsyntax_camltex.tex
1089
manual/src/refman/extensions/firstclassmodules.tex
1090
manual/src/refman/extensions/firstclassmodules_camltex.tex
1091
manual/src/refman/extensions/gadts.tex
1092
manual/src/refman/extensions/gadts_camltex.tex
1093
manual/src/refman/extensions/generalizedopens.tex
1094
manual/src/refman/extensions/generalizedopens_camltex.tex
1095
manual/src/refman/extensions/generativefunctors.tex
1096
manual/src/refman/extensions/generativefunctors_camltex.tex
1097
manual/src/refman/extensions/indexops.tex
1098
manual/src/refman/extensions/indexops_camltex.tex
1099
manual/src/refman/extensions/inlinerecords.tex
1100
manual/src/refman/extensions/inlinerecords_camltex.tex
1101
manual/src/refman/extensions/letrecvalues.tex
1102
manual/src/refman/extensions/letrecvalues_camltex.tex
1103
manual/src/refman/extensions/locallyabstract.tex
1104
manual/src/refman/extensions/locallyabstract_camltex.tex
1105
manual/src/refman/extensions/modulealias.tex
1106
manual/src/refman/extensions/modulealias_camltex.tex
1107
manual/src/refman/extensions/moduletypeof.tex
1108
manual/src/refman/extensions/moduletypeof_camltex.tex
1109
manual/src/refman/extensions/overridingopen.tex
1110
manual/src/refman/extensions/overridingopen_camltex.tex
1111
manual/src/refman/extensions/recursivemodules.tex
1112
manual/src/refman/extensions/recursivemodules_camltex.tex
1113
manual/src/refman/extensions/signaturesubstitution.tex
1114
manual/src/refman/extensions/signaturesubstitution_camltex.tex'
1115
Error: Process completed with exit code 1.
```

**Comment by Octachron** [comment]
You can extend the gitignore file in `refman` to also ignore the `tex` files in `refman/extensions` .

**Comment by johnwhitington** [comment]
`git check-ignore` thinks my `.gitignore` file is fine:

```
$ git check-ignore -v extensions/*.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/alerts.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/alerts_camltex.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/attributes.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/attributes_camltex.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/bigarray.tex
manual/src/refman/.gitignore:3:/extensions/*.tex	extensions/bigarray_camltex.tex
(many more lines)
```

But the flambda build still fails in the same way. The `.gitignore` is:

```
*.tex
*.htex
/extensions/*.tex
/extensions/*.htex
```

**Comment by Octachron** [comment]
I was mistaken, `git ls-files --others -i --exclude-standard` is checking that all non-tracked files are erased by the clean rule, and the clean rule has not been updated for `extensions`.

**Comment by gasche** [comment]
This sounds like a test whose output should be clarified with non-expert advice on what it checks how to fix it. @Octachron or @lthls, could either of you take care of this?

**Comment by johnwhitington** [comment]
Ok, we're finally ready. Thanks everyone.

**Comment by Octachron** [comment]
There is a conflict, which is seems to be due to presence of unrelated commits in this PR (that also make hard to check the final  state amongst the unrelated changes). It might be that you have rebased your branch on an out-of-date local `trunk`.

Could you try to rebase this PR in three steps:
- first update your local trunk branch with `git checkout trunk && git pull --ff-only name_of_upstream_remote trunk`
- rebase your branch with `git checkout split-exten-dot-etex && git rebase trunk`
- force push the updated branch with `git push -f name_of_github_remote split-exten-dot-etex`

**Comment by alainfrisch** [comment]
> I will have proposals on (2) later.

My proposal would be to drop the concept of "Language extensions", which really has no meaning in my opinion.  New features should be moved to the part of the manual where they really belong, perhaps with an indication of the OCaml version where they have been introduced (as we do for stdlib additions),  and possibly with some disclaimer about them being still considered as experimental, if applicable (but that should be rare).

**Comment by johnwhitington** [comment]
I am beginning to regret changing `manual/manual` to `manual/src` :-)

Your suggested rebase fails with:

```
Falling back to patching base and 3-way merge...
Auto-merging manual/src/tutorials/Makefile
CONFLICT (content): Merge conflict in manual/src/tutorials/Makefile
CONFLICT (modify/delete): manual/manual/refman/exten.etex deleted in HEAD and modified in Finished split of exten.etex. Version Finished split of exten.etex of manual/manual/refman/exten.etex left in tree.
error: Failed to merge in the changes.
Patch failed at 0008 Finished split of exten.etex
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

gorge:ocaml john$ git am --show-current-patch
commit 0b41e5d00ef3f1760300ffc727bbc1ab576e5dac (origin/gadt-tutorial)
Author: John Whitington <john@coherentgraphics.co.uk>
Date:   Tue Jan 19 18:55:39 2021 +0000

    Finished split of exten.etex
```
(the failed patch appears to be almost all the changes)

If we cannot rescue this, I could manually prepare two completely fresh PRs: one to split the files, and then another to move the files. Let me know.

**Comment by Octachron** [comment]
Here is a rebased branch: https://github.com/Octachron/ocaml/commits/pr10165bis . If this looks fine with you, I can (force) push the branch to update the PR branch.

**Comment by johnwhitington** [comment]
Thanks. I have downloaded and built your rebased branch, and the manual builds correctly, and all the changes intended appear to have been made. I also had a look over the list of commits, and they all seem to be correct.

**Review by Octachron — APPROVED** [review]
Splitting the extension file seems like a good idea. My muscle memory will take time to adapt to `manual/src`, but it seems clearer for people that don't have such memory.

I will merge once the CI is green to avoid further rebasing troubles.

**Comment by gasche** [comment]
@Octachron should one squash or merge?

**Comment by Octachron** [comment]
There is some smaller fix commits, but I would prefer to not squash together the splitting of `exten.etex` and the renaming commits. It thus seems better to merge rather than squash.

**Comment by gasche** [comment]
There is an ironical CI failure:

```diff
# ------------------------------
# This test checks that running tools/sync-stdlib-docs is a no-op in the current
# state, which means that the labelled/unlabelled .mli files are in sync.  If
# this check fails, it should be fixable by just running the script and reviewing
# the changes it makes.
# ------------------------------------------------------------------------
# +tools/sync_stdlib_docs
# +git diff --quiet --exit-code
# +result=fail
# +case $result in
# +echo 'CheckSyncStdlibDocs: failure with the following differences:'
# CheckSyncStdlibDocs: failure with the following differences:
# +git --no-pager diff
diff --git a/otherlibs/unix/unix.mli b/otherlibs/unix/unix.mli
index 1a22419..f4da667 100644
--- a/otherlibs/unix/unix.mli
+++ b/otherlibs/unix/unix.mli
@@ -25,7 +25,7 @@
    When a new function is added which is not implemented on Windows (or
    partially implemented), or the Windows-status of an existing function is
    changed, remember to update the summary table in
-   manual/manual/library/libunix.etex
+   manual/src/library/libunix.etex
 *)
```

**Comment by johnwhitington** [comment]
Done. Travis failure is just Changes.

**Pull request merged**